hw/ip_templates/ac_range_check/rtl/ac_range_check.sv.tpl

```diff
@@ -76,6 +79,210 @@ module ${module_instance_name}
+    .tl_h2d_o      ( ctn_filtered_tl_h2d_o ),
+    .tl_d2h_i      ( ctn_filtered_tl_d2h_i )
+  );
+
+  //////////////////////////////////////////////////////////////////////////////
+  // Range Check Deny Counting Logic
+  //////////////////////////////////////////////////////////////////////////////
+
+  logic [DenyCountWidth-1:0] deny_cnt;
+  logic deny_cnt_incr, deny_cnt_clr;
+
+  // Only increment the deny counter if logging is enabled
+  assign deny_cnt_incr = reg2hw.log_config.log_enable.q & range_check_fail;
+  // Determine if we are doing the first log. This one is special, since it also needs to log
+  // diagnostics data
+  assign log_first_deny = deny_cnt_incr & (deny_cnt == 0);
+
+  // Clear log information when clearing the interrupt or when clearing the log manually via the
+  // the log_clear bit.
+  logic intr_state_cleared, clear_log;
+  assign clear_log = intr_state_cleared |
+                     (reg2hw.log_config.log_clear.qe & reg2hw.log_config.log_clear.q);
+
+  prim_count #(
+    .Width(DenyCountWidth)
+  ) u_deny_count (
+    .clk_i              ( clk_i              ),
+    .rst_ni             ( rst_ni             ),
+    .clr_i              ( clear_log          ),
+    .set_i              ( 1'b0               ),
+    .set_cnt_i          ( '0                 ),
+    .incr_en_i          ( deny_cnt_incr      ),
+    .decr_en_i          ( 1'b0               ),
+    .step_i             ( DenyCountWidth'(1) ),
+    .commit_i           ( 1'b1               ),
+    .cnt_o              ( deny_cnt           ),
+    .cnt_after_commit_o (                    ),
+    .err_o              ( deny_cnt_error     )
+  );
+
+  // Log count is transparently mirrored. Clearing happens on the counter.
+  assign hw2reg.log_status.deny_count.de = 1'b1;
+  assign hw2reg.log_status.deny_count.d  = deny_cnt;
+
+  assign hw2reg.log_status.denied_read_access.de = log_first_deny | clear_log;
+  assign hw2reg.log_status.denied_read_access.d  = log_first_deny ? read_access : 1'b0;
+
+  assign hw2reg.log_status.denied_write_access.de = log_first_deny | clear_log;
+  assign hw2reg.log_status.denied_write_access.d  = log_first_deny ? write_access : 1'b0;
+
+  assign hw2reg.log_status.denied_execute_access.de = log_first_deny | clear_log;
+  assign hw2reg.log_status.denied_execute_access.d  = log_first_deny ? execute_access : 1'b0;
+
+  assign hw2reg.log_status.denied_no_match.de = log_first_deny | clear_log;
+
+  // Identify if the request was denied because there was no range matching at all. This occurs
+  // when the corresponding access mask is 0, meaning there was no address hit at all.
+  case unique ({read_access, write_access, exec_access})
+    3'b1??:  hw2reg.log_status.denied_no_match.d = read_mask    == '0;
+    3'b?1?:  hw2reg.log_status.denied_no_match.d = write_mask   == '0;
+    3'b??1:  hw2reg.log_status.denied_no_match.d = execute_mask == '0;
+    default: hw2reg.log_status.denied_no_match.d = 1'b0;
+  endcase
```

Comment:
Wouldn't it be much simpler to pull `addr_hit` out of `gen_range_checks` and make it an array over `NumRanges`, and then `assign hw2reg.log_status.denied_no_match.d = ~(|addr_hit)`?

Reply:
I guess it's ok. It would count all hits regardless of the permissions.